File: API/AmusementPark.WebAPI/Contracts/Videos/VideoDto.cs
using AmusementPark.WebAPI.Contracts.Common;

namespace AmusementPark.WebAPI.Contracts.Videos;

public sealed class VideoDto
{
    public string Id { get; set; } = string.Empty;

    public VideoHostingProviderDto HostingProvider { get; set; }

    public VideoOwnerTypeDto OwnerType { get; set; }

    public string? OwnerId { get; set; }

    public VideoTypeDto Type { get; set; }

    public string OriginalUrl { get; set; } = string.Empty;

    public string CanonicalUrl { get; set; } = string.Empty;

    public string? EmbedUrl { get; set; }

    public string? ExternalId { get; set; }

    public string Title { get; set; } = string.Empty;

    public string? Description { get; set; }

    public string? CreatorName { get; set; }

    public string? CreatorUrl { get; set; }

    public string? ThumbnailUrl { get; set; }

    public string? ThumbnailImageId { get; set; }

    public long? DurationSeconds { get; set; }

    public DateTime? PublishedAtUtc { get; set; }

    public List<LocalizedTextDto> Titles { get; set; } = new();

    public List<LocalizedTextDto> Descriptions { get; set; } = new();

    public List<string> TagIds { get; set; } = new();

    public VideoExternalMetadataDto ExternalMetadata { get; set; } = new();

    public bool IsPublished { get; set; }

    public DateTime CreatedAt { get; set; }

    public DateTime UpdatedAt { get; set; }
}

public sealed class VideoExternalMetadataDto
{
    public string? Source { get; set; }

    public DateTime? FetchedAtUtc { get; set; }

    public string? ProviderTitle { get; set; }

    public string? ProviderDescription { get; set; }

    public string? ProviderChannelId { get; set; }

    public string? ProviderChannelUrl { get; set; }
}

File: FRONT/AmusementPark/src/app/models/videos/video-dto.ts
import { LocalizedItemDto } from '../shared/localized-item-dto';
import { VideoHostingProvider } from './video-hosting-provider';
import { VideoOwnerType } from './video-owner-type';
import { VideoType } from './video-type';

export interface VideoDto {
  id: string;
  hostingProvider: VideoHostingProvider;
  ownerType: VideoOwnerType;
  ownerId?: string | null;
  type: VideoType;
  originalUrl: string;
  canonicalUrl: string;
  embedUrl?: string | null;
  externalId?: string | null;
  title: string;
  description?: string | null;
  creatorName?: string | null;
  creatorUrl?: string | null;
  thumbnailUrl?: string | null;
  thumbnailImageId?: string | null;
  durationSeconds?: number | null;
  publishedAtUtc?: string | null;
  languageCodes: string[];
  titles: LocalizedItemDto<string>[];
  descriptions: LocalizedItemDto<string>[];
  tagIds: string[];
  externalMetadata: VideoExternalMetadataDto;
  isPublished: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface VideoExternalMetadataDto {
  source?: string | null;
  fetchedAtUtc?: string | null;
  providerTitle?: string | null;
  providerDescription?: string | null;
  providerChannelId?: string | null;
  providerChannelUrl?: string | null;
}
